File: src/agent/messages.ts
export interface ToolCall {
  id: string;
  name: string;
  arguments: unknown;
}

export interface UserMessage {
  role: "user";
  content: string;
}

export interface AssistantMessage {
  role: "assistant";
  content: string;
  toolCalls: ToolCall[];
  providerMetadata?: Record<string, unknown>;
}

export interface ToolMessage {
  role: "tool";
  toolCallId: string;
  toolName: string;
  ok: boolean;
  content: string;
  data?: unknown;
  error?: ToolError;
}

export interface ToolError {
  code: string;
  message: string;
}

export type Message = UserMessage | AssistantMessage | ToolMessage;

File: src/forge_agent/messages.py
"""Provider-neutral transcript message models."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from forge_agent.tools import ToolCall
from forge_agent.types import JSONValue


class UserMessage(BaseModel):
    """A message authored by the user."""

    model_config = ConfigDict(extra="forbid")

    role: Literal["user"] = "user"
    content: str


class AssistantMessage(BaseModel):
    """A message authored by the assistant, optionally requesting tool calls."""

    model_config = ConfigDict(extra="forbid")

    role: Literal["assistant"] = "assistant"
    content: str = ""
    tool_calls: list[ToolCall] = Field(default_factory=list)


class ToolResultMessage(BaseModel):
    """A transcript message containing the result of a previous tool call."""

    model_config = ConfigDict(extra="forbid")

    role: Literal["tool"] = "tool"
    tool_call_id: str
    name: str
    content: str
    ok: bool = True
    data: dict[str, JSONValue] | None = None
    details: dict[str, JSONValue] | None = None
    error: str | None = None


type AgentMessage = UserMessage | AssistantMessage | ToolResultMessage
